### DISCUSSION TITLE
Xstate performance A >x10 factor

### DISCUSSION THREAD

**Discussion by lveillard** [statelyai/xstate#4747] [Q&A]
Hello! I have a machine with 6-7, linear right now so nothing complex. It runs automatically, just one state after the other once the requests are resolved.

Invoking all the actors in a series takes about 10ms but when i run them in Xstate, it takes about 160ms.

Is this normal?
Any idea how to find out what is adding those delays?

Thanks!

**Comment by davidkpiano** [comment]
Can you share the code for this?

**Reply by lveillard** [reply]
Sorry, missed the notification.
So we migrated our old pipeline here (https://github.com/Blitzapps/blitz-orm/blob/main/src/pipeline/pipeline.ts:runPipeline)
To a machine that looks like this one: https://github.com/Blitzapps/blitz-orm/blob/main/src/stateMachine/query/machine.ts

I'm sorry i no longer have the xstate example. Because of the perf drop we moved to robot3.

For context, it is an ORM that does a query/mutation to a graph database (typeDB) in multiple steps. It prepares it, it runs the query, it formats it output etc.  A > B > C ... Obviously some of these operations are async

The old pipeline (runPipeline fn) is obviously the faster.  Then robot3 was slower but kinda ok and finaly xState which, for some reason was up to 10 times slower than the runPipeline fn.
